pkg/sidecar/sidecar.go

```diff
@@ -17,59 +17,96 @@ import (
 	"syscall"
 	"time"
 
+	"github.com/andres-erbsen/clock"
+	proto "github.com/spiffe/go-spiffe/proto/spiffe/workload"
 	"github.com/spiffe/spire/api/workload"
-	proto "github.com/spiffe/spire/proto/api/workload"
+	"golang.org/x/sys/unix"
 )
 
-// sidecar is the component that consumes the Workload API and renews certs
+// Config contains config variables when creating a SPIFFE Sidecar.
+type Config struct {
+	AgentAddress       string
+	Cmd                string
+	CmdArgs            string
+	CertDir            string
+	SvidFileName       string
+	SvidKeyFileName    string
+	SvidBundleFileName string
+	RenewSignal        string
+	Timeout            string
+	ExternalProcess    bool
+	ProcessReloadFunc  func() error
+}
+
+// Sidecar is the component that consumes the Workload API and renews certs
 // implements the interface Sidecar
-type sidecar struct {
-	config            *SidecarConfig
+type Sidecar struct {
+	config            *Config
 	processRunning    int32
 	process           *os.Process
 	workloadAPIClient workload.X509Client
+	certReadyChan     chan struct{}
 }
 
 const (
 	// default timeout Duration for the workloadAPI client when the defaultTimeout
 	// is not configured in the .conf file
-	defaultTimeout = time.Duration(5 * time.Second)
+	defaultTimeout = 5 * time.Second
+	delayMin       = time.Second
+	delayMax       = time.Minute
 
 	certsFileMode = os.FileMode(0644)
 	keyFileMode   = os.FileMode(0600)
 )
 
-// NewSidecar creates a new sidecar
-func NewSidecar(config *SidecarConfig) (*sidecar, error) {
+// NewSidecar creates a new SPIFFE sidecar
+func NewSidecar(config *Config) (*Sidecar, error) {
 	timeout, err := getTimeout(config)
 	if err != nil {
 		return nil, err
 	}
 
-	return &sidecar{
+	return &Sidecar{
 		config:            config,
 		workloadAPIClient: newWorkloadAPIClient(config.AgentAddress, timeout),
+		certReadyChan:     make(chan struct{}, 1),
 	}, nil
 }
 
 // RunDaemon starts the main loop
 // Starts the workload API client to listen for new SVID updates
 // When a new SVID is received on the updateChan, the SVID certificates
 // are stored in disk and a restart signal is sent to the proxy's process
-func (s *sidecar) RunDaemon(ctx context.Context) error {
+func (s *Sidecar) RunDaemon(ctx context.Context) {
 	// Create channel for interrupt signal
 	interrupt := make(chan os.Signal, 1)
 	errorChan := make(chan error, 1)
 	signal.Notify(interrupt, syscall.SIGINT, syscall.SIGTERM)
 
 	updateChan := s.workloadAPIClient.UpdateChan()
 
-	//start the workloadAPIClient
+	// start the workloadAPIClient
 	go func() {
-		err := s.workloadAPIClient.Start()
-		if err != nil {
-			log.Println(err.Error())
-			errorChan <- err
+		clk := clock.New()
+		delay := delayMin
+		for {
+			err := s.workloadAPIClient.Start()
+			if err != nil {
```

Comment:
What problem was this change trying to solve? As configured, Start() won't return until Stop has been called. The client already takes care of backoff and executing retries internally...

Reply:
I didnt change this section of the code so Im not sure why its showing up here. This code exists in master.

Reply:
Oh weird. I think your chain of commits has some of the other commits that were merged into master in there. I'm not sure if they were cherry-picked or something but don't seem have been introduced into your branch via a merge commit so they show up as explicit commits in your PR, which is why github is showing it the way that it is.

We're pretty close on this and there are some merge conflicts with master. Maybe the right thing to do is to rebase this on top of master, fix up the conflicts and squash it all down to a single commit? Then we're not trying to review code that you didn't introduce...

Reply:
This is done.